src/flag_gems/ops/unique.py

```diff
@@ -0,0 +1,846 @@
+            output_counts_flat_kernel[grid](
+                idx,
+                num_tasks,  # in
+                counts,  # out
+                out_size,
+                tiles_per_cta,
+                tile_size,
+                one_tile_per_cta=tiles_per_cta == 1,
+                num_warps=num_warps,
+            )
+
+    return data_out[:out_size], inverse_indices, counts
+
+
+def simple_unique_flat(
+    sorted_data: torch.Tensor,
+    sorted_indices: torch.Tensor,
+    return_inverse: bool,
+    return_counts: bool,
+):
+    num_tasks = sorted_data.numel()
+    grid = (1, 1, 1)
+
+    # allocate tensor
+    data_out = torch.empty_like(sorted_data)
+    if return_inverse:
+        inverse_indices = torch.empty_like(sorted_data, dtype=torch.int32)
+    else:
+        inverse_indices = None
+    if return_counts:
+        idx = torch.empty_like(sorted_data, dtype=torch.int32)
+    else:
+        idx = None
+    unique_size = torch.empty([1], dtype=torch.int32, device=sorted_data.device)
+
+    # launch kernel
+    with torch.cuda.device(sorted_data.device.index):
+        simple_unique_flat_kernel[grid](
+            sorted_data,
+            sorted_indices,  # in
+            data_out,
+            inverse_indices,
+            idx,
+            unique_size,  # out
+            return_inverse,
+            return_counts,
+            num_tasks,
+            tile_size=triton.next_power_of_2(num_tasks),
+            num_warps=8,
+        )
+    out_size = unique_size.item() + 1
+    counts = None
+    if return_counts:
+        idx = idx[:out_size]
+        counts = torch.empty_like(idx)
+        with torch.cuda.device(sorted_data.device.index):
+            output_counts_flat_kernel[grid](
+                idx,
+                num_tasks,  # in
+                counts,  # out
+                num_tasks=out_size,
+                tiles_per_cta=1,
+                tile_size=triton.next_power_of_2(out_size),
+                one_tile_per_cta=True,
+                num_warps=8,
+            )
+    return data_out[:out_size], inverse_indices, counts
+
+
+def unique_flat(
+    in0: torch.Tensor,
+    sorted: bool = True,
+    return_inverse: bool = False,
+    return_counts: bool = False,
+):
+    if in0.numel() <= 8192:
+        sorted_data, sorted_indices = torch.sort(in0.ravel(), stable=False)
+        data_out, inverse_indices, counts = simple_unique_flat(
+            sorted_data, sorted_indices, return_inverse, return_counts
+        )
+    elif return_inverse:
+        sorted_data, sorted_indices = torch.sort(in0.ravel(), stable=False)
+        data_out, inverse_indices, counts = sorted_indices_unique_flat(
+            sorted_data, sorted_indices, return_counts
+        )
+    else:
+        sorted_data, _ = torch.sort(in0.ravel(), stable=False)
+        data_out, inverse_indices, counts = sorted_quick_unique_flat(
+            sorted_data, return_counts
+        )
+    return data_out, inverse_indices, counts
+
+
+def unique(
```

Comment:
> Also, I don't see why we need a redirected call to unique_flat. Their function prototypes are literally the same.

Because isin depends on unique_flat, I initially split it into two functions. But now that view_as is not expensive, I will merge it into one function.

Reply:
I see.